File: front-end/src/domain/value-object/x-position.ts
export class XPosition {
    private static readonly ERROR_X_POSITION_INVALID = 'X position must be a valid number (not NaN)';

    private readonly xPosition: number;

    constructor(xPosition: number) {
        XPosition.validate(xPosition);
        this.xPosition = xPosition;
    }

    private static validate(xPosition: number): void {
        if (Number.isNaN(xPosition)) {
            throw new Error(XPosition.ERROR_X_POSITION_INVALID);
        }
    }

    public getXPosition(): number {
        return this.xPosition;
    }

    public setXPosition(xPosition: number): XPosition {
        return new XPosition(xPosition);
    }

    public toString(): string {
        return this.xPosition.toString();
    }

    public equals(other: unknown): boolean {
        return (
            other instanceof XPosition &&
            this.xPosition === other.xPosition
        );
    }

    public hashCode(): number {
        const buffer = new ArrayBuffer(8);
        new Float64Array(buffer)[0] = this.xPosition;
        return new Int32Array(buffer)[0];
    }
}

File: back-end/src/main/java/es/ull/project/domain/valueobject/XPosition.java
package es.ull.project.domain.valueobject;

import java.math.BigDecimal;
import java.util.Objects;

/**
 * @brief Represents the X position of a node in a workflow.
 */
public class XPosition {

    private static final String ERROR_X_POSITION_INVALID = "X position must be a valid non-null BigDecimal";
    private final BigDecimal position;

    public XPosition(BigDecimal xPosition) {
        this.validate(xPosition);
        this.position = xPosition;
    }

    private void validate(BigDecimal xPosition) {
        if (xPosition == null) {
            throw new IllegalArgumentException(ERROR_X_POSITION_INVALID);
        }
    }

    public BigDecimal getXPosition() {
        return this.position;
    }

    public XPosition setXPosition(BigDecimal xPosition) {
        return new XPosition(xPosition);
    }

    @Override
    public boolean equals(Object otherObject) {
        if (this == otherObject) {
            return true;
        }
        if (!(otherObject instanceof XPosition)) {
            return false;
        }
        final XPosition that = (XPosition) otherObject;
        return this.position.compareTo(that.position) == 0;
    }

    @Override
    public int hashCode() {
        return Objects.hash(this.position);
    }

    @Override
    public String toString() {
        return this.position.toString();
    }
}